File: core/src/telemetry/telemetry.cpp
// SPDX-License-Identifier: MIT
#include <memogent/telemetry.hpp>

#include <nlohmann/json.hpp>

#include <algorithm>
#include <deque>
#include <mutex>
#include <string>
#include <unordered_map>

namespace mem {

namespace {

using json = nlohmann::json;

class InMemoryTelemetry final : public Telemetry {
public:
    explicit InMemoryTelemetry(const Config& cfg) : cfg_(cfg) {}

    void record_event(std::string_view kind,
                      const std::unordered_map<std::string, std::string>& attrs) override {
        std::lock_guard g(mu_);
        json j = {{"kind", std::string{kind}}};
        for (const auto& [k, v] : attrs) j[k] = v;
        events_.push_back(j.dump());
        trim();
    }

    void record_latency(std::string_view kind, double ms) override {
        std::lock_guard g(mu_);
        latencies_[std::string{kind}].push_back(ms);
    }

    void record_prediction(bool top1, bool top3) override {
        std::lock_guard g(mu_);
        ++predictions_total_;
        if (top1) ++predictions_top1_;
        if (top3) ++predictions_top3_;
    }

    void record_cache(std::string_view pool, std::size_t hits, std::size_t misses) override {
        std::lock_guard g(mu_);
        auto& c = cache_[std::string{pool}];
        c.first = hits; c.second = misses;
    }

    void record_thrash(std::size_t events) override {
        std::lock_guard g(mu_); thrash_events_ += events;
    }

    void record_stability_issue(std::string_view reason) override {
        std::lock_guard g(mu_);
        ++stability_issues_;
        stability_reasons_.emplace_back(reason);
    }

    KpiSnapshot snapshot() const override {
        std::lock_guard g(mu_);
        KpiSnapshot k{};
        k.prediction_accuracy_top1 = predictions_total_ == 0
            ? 0.0 : static_cast<double>(predictions_top1_) / predictions_total_;
        k.prediction_accuracy_top3 = predictions_total_ == 0
            ? 0.0 : static_cast<double>(predictions_top3_) / predictions_total_;

        std::size_t total_h = 0, total_m = 0;
        for (const auto& [_, v] : cache_) { total_h += v.first; total_m += v.second; }
        k.cache_hit_rate = (total_h + total_m) == 0 ? 0.0
            : static_cast<double>(total_h) / (total_h + total_m);

        k.stability_issues = stability_issues_;
        k.thrashing_events_current = static_cast<double>(thrash_events_);
        k.thrashing_events_baseline = baseline_thrashing_;
        k.thrashing_reduction_pct = baseline_thrashing_ == 0 ? 0.0
            : (baseline_thrashing_ - thrash_events_) * 100.0 / baseline_thrashing_;

        auto percentile = [](std::vector<double> v, double p) {
            if (v.empty()) return 0.0;
            std::sort(v.begin(), v.end());
            const auto idx = std::min<std::size_t>(v.size() - 1,
                                                   static_cast<std::size_t>(v.size() * p));
            return v[idx];
        };
        std::vector<double> decisions;
        if (auto it = latencies_.find("decision"); it != latencies_.end())
            decisions = it->second;
        k.p50_decision_latency_ms = percentile(decisions, 0.50);
        k.p95_decision_latency_ms = percentile(decisions, 0.95);
        k.p99_decision_latency_ms = percentile(decisions, 0.99);

        k.app_load_time_ms_baseline = baseline_app_load_ms_;
        k.app_load_time_ms_current = current_app_load_ms_;
        k.app_load_time_improvement_pct = baseline_app_load_ms_ == 0.0 ? 0.0
            : (baseline_app_load_ms_ - current_app_load_ms_) * 100.0 / baseline_app_load_ms_;
        k.launch_time_ms_baseline = baseline_launch_ms_;
        k.launch_time_ms_current = current_launch_ms_;
        k.launch_time_improvement_pct = baseline_launch_ms_ == 0.0 ? 0.0
            : (baseline_launch_ms_ - current_launch_ms_) * 100.0 / baseline_launch_ms_;
        k.memory_utilization_efficiency_pct = util_efficiency_;
        return k;
    }

    std::string to_json() const override {
        const auto k = snapshot();
        json j = {
            {"app_load_time_ms_baseline", k.app_load_time_ms_baseline},
            {"app_load_time_ms_current", k.app_load_time_ms_current},
            {"app_load_time_improvement_pct", k.app_load_time_improvement_pct},
            {"launch_time_ms_baseline", k.launch_time_ms_baseline},
            {"launch_time_ms_current", k.launch_time_ms_current},
            {"launch_time_improvement_pct", k.launch_time_improvement_pct},
            {"thrashing_events_baseline", k.thrashing_events_baseline},
            {"thrashing_events_current", k.thrashing_events_current},
            {"thrashing_reduction_pct", k.thrashing_reduction_pct},
            {"stability_issues", k.stability_issues},
            {"prediction_accuracy_top1", k.prediction_accuracy_top1},
            {"prediction_accuracy_top3", k.prediction_accuracy_top3},
            {"cache_hit_rate", k.cache_hit_rate},
            {"memory_utilization_efficiency_pct", k.memory_utilization_efficiency_pct},
            {"p50_decision_latency_ms", k.p50_decision_latency_ms},
            {"p95_decision_latency_ms", k.p95_decision_latency_ms},
            {"p99_decision_latency_ms", k.p99_decision_latency_ms},
        };
        return j.dump(2);
    }

    // These setters let the benchmark harness feed baseline numbers in.
    void set_baseline_app_load(double ms) { std::lock_guard g(mu_); baseline_app_load_ms_ = ms; }
    void set_current_app_load(double ms) { std::lock_guard g(mu_); current_app_load_ms_ = ms; }
    void set_baseline_launch(double ms) { std::lock_guard g(mu_); baseline_launch_ms_ = ms; }
    void set_current_launch(double ms) { std::lock_guard g(mu_); current_launch_ms_ = ms; }
    void set_util_efficiency(double pct) { std::lock_guard g(mu_); util_efficiency_ = pct; }
    void set_baseline_thrashing(double t) { std::lock_guard g(mu_); baseline_thrashing_ = t; }

private:
    void trim() {
        while (events_.size() > cfg_.telemetry_buffer) events_.pop_front();
    }

    mutable std::mutex mu_;
    Config cfg_;
    std::deque<std::string> events_;
    std::unordered_map<std::string, std::vector<double>> latencies_;
    std::size_t predictions_total_ = 0, predictions_top1_ = 0, predictions_top3_ = 0;
    std::unordered_map<std::string, std::pair<std::size_t, std::size_t>> cache_;
    std::size_t thrash_events_ = 0, stability_issues_ = 0;
    std::vector<std::string> stability_reasons_;
    double baseline_app_load_ms_ = 0, current_app_load_ms_ = 0;
    double baseline_launch_ms_ = 0, current_launch_ms_ = 0;
    double util_efficiency_ = 0;
    double baseline_thrashing_ = 0;
};

}  // namespace

Result<std::unique_ptr<Telemetry>> make_telemetry(const Config& cfg) {
    return std::unique_ptr<Telemetry>(new InMemoryTelemetry(cfg));
}

}  // namespace mem

File: python/memogent/telemetry.py
"""In-memory Telemetry — mirrors core/src/telemetry/telemetry.cpp.

Accumulates every Samsung KPI and exposes a snapshot + JSON dump.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Dict, List


@dataclass
class KpiSnapshot:
    app_load_time_ms_baseline: float = 0.0
    app_load_time_ms_current: float = 0.0
    app_load_time_improvement_pct: float = 0.0
    launch_time_ms_baseline: float = 0.0
    launch_time_ms_current: float = 0.0
    launch_time_improvement_pct: float = 0.0
    thrashing_events_baseline: float = 0.0
    thrashing_events_current: float = 0.0
    thrashing_reduction_pct: float = 0.0
    stability_issues: int = 0
    prediction_accuracy_top1: float = 0.0
    prediction_accuracy_top3: float = 0.0
    cache_hit_rate: float = 0.0
    memory_utilization_efficiency_pct: float = 0.0
    p50_decision_latency_ms: float = 0.0
    p95_decision_latency_ms: float = 0.0
    p99_decision_latency_ms: float = 0.0


@dataclass
class Telemetry:
    baseline_app_load_ms: float = 0.0
    current_app_load_ms: float = 0.0
    baseline_launch_ms: float = 0.0
    current_launch_ms: float = 0.0
    baseline_thrashing: float = 0.0
    current_thrashing: float = 0.0
    util_efficiency_pct: float = 0.0
    stability_issues: int = 0
    _pred_total: int = 0
    _pred_top1: int = 0
    _pred_top3: int = 0
    _caches: Dict[str, tuple[int, int]] = field(default_factory=dict)
    _latencies: Dict[str, List[float]] = field(default_factory=dict)

    # ---- mutators ----
    def record_prediction(self, top1: bool, top3: bool) -> None:
        self._pred_total += 1
        self._pred_top1 += int(top1)
        self._pred_top3 += int(top3)

    def record_cache(self, pool: str, hits: int, misses: int) -> None:
        self._caches[pool] = (hits, misses)

    def record_latency(self, kind: str, ms: float) -> None:
        self._latencies.setdefault(kind, []).append(ms)

    def record_stability_issue(self, _reason: str) -> None:
        self.stability_issues += 1

    def record_thrash(self, events: int) -> None:
        self.current_thrashing += events

    # ---- snapshot ----
    def snapshot(self) -> KpiSnapshot:
        k = KpiSnapshot()
        k.prediction_accuracy_top1 = (
            self._pred_top1 / self._pred_total if self._pred_total else 0.0
        )
        k.prediction_accuracy_top3 = (
            self._pred_top3 / self._pred_total if self._pred_total else 0.0
        )
        total_h = sum(h for h, _ in self._caches.values())
        total_m = sum(m for _, m in self._caches.values())
        k.cache_hit_rate = total_h / (total_h + total_m) if (total_h + total_m) else 0.0
        k.stability_issues = self.stability_issues
        k.thrashing_events_baseline = self.baseline_thrashing
        k.thrashing_events_current = self.current_thrashing
        k.thrashing_reduction_pct = (
            (self.baseline_thrashing - self.current_thrashing) * 100.0 / self.baseline_thrashing
            if self.baseline_thrashing
            else 0.0
        )
        k.app_load_time_ms_baseline = self.baseline_app_load_ms
        k.app_load_time_ms_current = self.current_app_load_ms
        if self.baseline_app_load_ms:
            k.app_load_time_improvement_pct = (
                (self.baseline_app_load_ms - self.current_app_load_ms)
                * 100.0
                / self.baseline_app_load_ms
            )
        k.launch_time_ms_baseline = self.baseline_launch_ms
        k.launch_time_ms_current = self.current_launch_ms
        if self.baseline_launch_ms:
            k.launch_time_improvement_pct = (
                (self.baseline_launch_ms - self.current_launch_ms)
                * 100.0
                / self.baseline_launch_ms
            )
        k.memory_utilization_efficiency_pct = self.util_efficiency_pct
        decisions = self._latencies.get("decision", [])
        if decisions:
            sorted_d = sorted(decisions)
            k.p50_decision_latency_ms = sorted_d[len(sorted_d) // 2]
            k.p95_decision_latency_ms = sorted_d[min(len(sorted_d) - 1, int(len(sorted_d) * 0.95))]
            k.p99_decision_latency_ms = sorted_d[min(len(sorted_d) - 1, int(len(sorted_d) * 0.99))]
        return k

    def to_dict(self) -> dict:
        s = self.snapshot()
        return {
            "app_load_time_ms_baseline": s.app_load_time_ms_baseline,
            "app_load_time_ms_current": s.app_load_time_ms_current,
            "app_load_time_improvement_pct": s.app_load_time_improvement_pct,
            "launch_time_ms_baseline": s.launch_time_ms_baseline,
            "launch_time_ms_current": s.launch_time_ms_current,
            "launch_time_improvement_pct": s.launch_time_improvement_pct,
            "thrashing_events_baseline": s.thrashing_events_baseline,
            "thrashing_events_current": s.thrashing_events_current,
            "thrashing_reduction_pct": s.thrashing_reduction_pct,
            "stability_issues": s.stability_issues,
            "prediction_accuracy_top1": s.prediction_accuracy_top1,
            "prediction_accuracy_top3": s.prediction_accuracy_top3,
            "cache_hit_rate": s.cache_hit_rate,
            "memory_utilization_efficiency_pct": s.memory_utilization_efficiency_pct,
            "p50_decision_latency_ms": s.p50_decision_latency_ms,
            "p95_decision_latency_ms": s.p95_decision_latency_ms,
            "p99_decision_latency_ms": s.p99_decision_latency_ms,
        }

    def to_json(self, indent: int = 2) -> str:
        return json.dumps(self.to_dict(), indent=indent)
